Make triggers work with the explicit table API

Nicolapps opened the pull request.

ianmacartney commented on packages/convex-helpers/server/triggers.ts:
```diff
@@ -168,43 +169,61 @@ export class DatabaseWriterWithTriggers<
-  ): Promise<void> {
-    return this.writer.patch(id, value);
+
+  patch(arg0: any, arg1: any, arg2?: any): Promise<void> {
```
do these `any` types show up in the customer's app? Or are they sufficiently hidden by the library otherwise?

Nicolapps replied:
Thanks, I fiexd the types in 875fe205321e000b157c58c75623a146e81075aa

Nicolapps commented on packages/convex-helpers/server/triggers.ts:
```diff
@@ -219,6 +238,73 @@ export function writerWithTriggers<
+  ): Promise<void> {
+    if (!tableName) {
+      return await innerDb.patch(id, value);
+    }
```
Yeah, the only reason why I’m doing that is to get the right types. Do you have a better idea?

Nicolapps commented:
> looking good so far - also remember to update streams & relationship helpers (and any other spots I'm forgetting)

@ianmacartney Thanks! I’m going to do a separate PR for row-level security. I don’t see custom implementations of `get`/`replace`/`patch`/`delete` in relationships/streams, is there something I’m missing?

Nicolapps commented:
> Looks good - is there some sanity check somewhere that it conforms to the DatabaseReader / Writer types? e.g. that ctx including a wrapped db will not have a type error? I guess the wrapDatabaseWriter controls that.

Yeah, since `writerWithTriggers` returns a `GenericDatabaseWriter` it is checked that way. The only part where type safety is worse is the methods that take `any`, for example:

```
patch: async (arg0: any, arg1: any, arg2?: any) => {
  const [tableName, id, value] =
    arg2 !== undefined
      ? [arg0, arg1, arg2]
      : [_tableNameFromId(innerDb, triggers.registered, arg0), arg0, arg1];
  return await _patch(tableName, id, value);
}
```

i.e. if we updated `patch` to take other arguments another day, it would still typecheck. I changed the implementation in 9adeba1c4e2421ffe1b2614186aa5dc1f756b5c6: by duplicating the type definition, we avoid this scenario. Let me know your thoughts on this!

----

> ah yeah stream only is on the queries, my mistake. And relationship helpers
> I think we can wait until it ships and either add deprecated overloads or
> deprecate some in general?

I see what you mean. We don’t need to change anything for now as both APIs are supported.
